### PR TITLE
Adding functionality to cordon the node before destroying it. 

### PR DESCRIPTION
**Pull request by atulaggarwal**
This helps load balancer to remove the node from healthy hosts (ALB does have this support).

This won't fix the issue of 502 completely as there is some time node to live even after cordoning as to serve In-Flight request.
Shamelessly copied from https://github.com/kubernetes/autoscaler/pull/3014/files

### CODE REVIEW

**Comment by atulaggarwal** [comment]
Signed now

**Comment by atulaggarwal** [comment]
Can someone review the PR and let me know if this kind of change make sense in autoscaler?

**Comment by MaciekPytel** [comment]
FWIW I still believe the right solution is to make ingress controller aware of autoscaler taint (ex. [ingress-gce](https://github.com/kubernetes/ingress-gce/issues/595)). spec.Unschedulable has the problem of not having a clear owner, though the idea of pairing it with a taint as implemented in here somewhat mitigates this issue.

At a glance I'm not sure why we need 2 separate taints? Why not just skip default tainting logic and apply the ToBeDeletedTaint in cordoning logic? Having 2 separate writes on the node is a big downside as mutating api-server calls can be a big bottleneck when trying to scale-down large cluster rapidly.

I'm focusing on preparing patch releases now, I can do a more detailed review later this week.

**Comment by atulaggarwal** [comment]
Thanks for the brief review. 
1. I am raising PR with aws lonader balancer to handle the taint.
2. I agree we should consolidate the taints at one place. I will work on it and make the changes. I still believe we should use spec.Unschedulable also so that standard implementatio unaware of autoscaler could listen for it.

**Comment by atulaggarwal** [comment]
@MaciekPytel - Can you review the PR once and let me know if everything looks fine?

**Comment by ltagliamonte-dd** [comment]
Folks why aren't we working on this? Can someone from the admin review?
this could easily solve a huge problem for tons of users.

**Review by feiskyer — COMMENTED** [review]
I'm +1 for adding this
/approve

**Comment by ltagliamonte-dd** [comment]
@feiskyer thank you for the review and the approval.
What are the remaining step to see this upstream and released?

**Comment by ltagliamonte-dd** [comment]
@Jeffwan do you mind review this PR as well?

File: cluster-autoscaler/utils/deletetaint/delete.go
Comment (mwielgus) [inline cluster-autoscaler/utils/deletetaint/delete.go:213]: Given that the update has not yet happened this message is premature.

File: cluster-autoscaler/main.go
Comment (mwielgus) [inline cluster-autoscaler/main.go:177]: Please launch new features initially disabled.

File: cluster-autoscaler/utils/deletetaint/delete.go
Comment (mwielgus) [inline cluster-autoscaler/utils/deletetaint/delete.go:121]: The update has not yet completed. The message is premature.

**Review by mwielgus — CHANGES_REQUESTED** [review]

**Comment by atulaggarwal** [comment]
@mwielgus - Thanks for reviewing the PR. I have made the changes as per the review comments. Please review it once more.

**Comment by mwielgus** [comment]
Please squash commits to just 1.

**Comment by atulaggarwal** [comment]
@mwielgus - Done

**Review by mwielgus — APPROVED** [review]
/lgtm
/approve

**Comment by MaciekPytel** [comment]
For future reference - there is a growing number of request for CA to take additional custom actions as part of the drain (ex. https://github.com/kubernetes/autoscaler/pull/3792) and I'd like to propose extracting drain logic into processor as a long-term solution.
This will keep any additional logic encapsulated and make it so that no code related to opt-in features will be executed if those features are disabled (which should make it easier to merge any such changes as any regression risk is limited to users who opt-in).

**Comment by ltagliamonte** [comment]
@mwielgus @feiskyer thank you for looking into this.
When can we expect a docker release?

**Comment by Jeffwan** [comment]
> @mwielgus @feiskyer thank you for looking into this.
> When can we expect a docker release?

There's a release cycle, you will need to cherry-pick this merge to branches and next release will pick them up. 
At the same time, you can build your own image to use and don't need to wait for official release if you want it immediately.

**Comment by ltagliamonte-dd** [comment]
@Jeffwan thank you for getting back to me, may you please tell me when the next release cycle is going to happen?

**Comment by spanky-medal** [comment]
Does this PR resolve the issue found in https://github.com/kubernetes/autoscaler/issues/2045? I'm not clear following this thread from https://github.com/kubernetes/autoscaler/pull/3792

**Comment by atulaggarwal** [comment]
I don't think this would solve issue #2045. This PR justs add UnSchedulable label to node being evicted so that it could be removed from load balancer. It would not change any way scheduling of pods before deleting the node.

**Comment by ltagliamonte-dd** [comment]
@Jeffwan i've created a dev docker release for my staging testing, still waiting on the official release for prod.
Do you know when it is going to happen? i'm waiting for a release on  k8s 1.18.
Thanks

**Comment by infa-ddeore** [comment]
@atulaggarwal I want to build a docker image for k8 `1.5`, `1.16`, `1.17`, `1.18` so trying to cherry-pick your commit `7670d7b6af6cf98e29f7f328d8de38072f7a0d9a` into the respective branches like `cluster-autoscaler-release-1.15`, `cluster-autoscaler-release-1.16` and so on, but there are conflicts in several files like `cluster-autoscaler/utils/deletetaint/delete.go` `cluster-autoscaler/main.go` `cluster-autoscaler/config/autoscaling_options.go`

would you be able to provide the fixes in other branches as well instead of master?

**Comment by Jeffwan** [comment]
> @Jeffwan i've created a dev docker release for my staging testing, still waiting on the official release for prod.
> Do you know when it is going to happen? i'm waiting for a release on k8s 1.18.
> Thanks

@ltagliamonte-dd Please cherry-pick this change to release branch. @kubernetes/autoscaler-maintainers will help cut release later.

**Pull request merged**
